src/Lower.cpp

```diff
@@ -332,6 +332,11 @@ Module lower(const vector<Function> &output_funcs,
     debug(2) << "Lowering after unrolling:\n"
              << s << "\n\n";
 
+    debug(1) << "Finding intrinsics...\n";
```

Comment:
Thanks for the suggestion. I noticed a failure in the simd_op_check when I ran it with the Sapphire Rapids target enabled after moving the find_intrinsics pass. I will investigate this today and report back.

Reply:
When I ran the tests in https://github.com/halide/Halide/pull/5807#discussion_r595841620 I forgot to enable the correct target to trigger the generation of the new instructions. After some experimentation I found this.

This first block is supposed to be `Lower.cpp` without modifications, so we have unrolling, vectorize and find_intrinsics. As you can see, find_intrinsics is able to replace the multiplication with a `widening_mul` and this is pattern expected in [CodeGen_X86](https://github.com/mcleary/Halide/blob/saturating_dot_products/src/CodeGen_X86.cpp#L542)

### Find Instrinsics After Vectorize Loops

```c++
// Unrolling
f[(f.s1.x.v1*16) + f.s1.x.v2] =
         let t7.s = int16((((((f.s1.x.v1*16) + f.min.0) + f.s1.x.v2)*2) + f.s1.rdom$x)) in
         saturating_add(f[(f.s1.x.v1*16) + f.s1.x.v2], int32(t7.s)*int32(t7.s))

// Vectorize
f[ramp(f.s1.x.v1*16, 1, 16) aligned(16, 0)] = (int32x16)saturating_add(f[ramp(f.s1.x.v1*16, 1, 16) aligned(16, 0)], (int32x16)vector_reduce(SaturatingAdd, (int32x32((int16x32)t7.s.widened.f.s1.rdom$x)*int32x32((int16x32)t7.s.widened.f.s1.rdom$x))))

// Find Intrinsics
f[ramp(f.s1.x.v1*16, 1, 16) aligned(16, 0)] = (int32x16)saturating_add(f[ramp(f.s1.x.v1*16, 1, 16) aligned(16, 0)], (int32x16)vector_reduce(SaturatingAdd, (int32x32)widening_mul((int16x32)t7.s.widened.f.s1.rdom$x, (int16x32)t7.s.widened.f.s1.rdom$x)))
```

When I placed `find_intrisics` before vectorize/unrolling I don't get the widening_mul in the expression, which won't match the expected pattern, thus, simd_op_check will fail for the expression

```c++
check("vpdpwssds*zmm", 16, saturating_sum(i32(0), i32(in_i16(2 * x + r)) * in_i16(2 * x + r + 32)));
```

## Find Intrinsics Before Vectorize Loops

```c++
// Unrolling
f[(f.s1.x.v1*16) + f.s1.x.v2] =
         let t7.s = int16((((((f.s1.x.v1*16) + f.min.0) + f.s1.x.v2)*2) + f.s1.rdom$x)) in
         saturating_add(f[(f.s1.x.v1*16) + f.s1.x.v2], int32(t7.s)*int32(t7.s))

// Find Intrisincs
f[(f.s1.x.v1*16) + f.s1.x.v2] =
         let t7.s = int16((((((f.s1.x.v1*16) + f.min.0) + f.s1.x.v2)*2) + f.s1.rdom$x)) in
         saturating_add(f[(f.s1.x.v1*16) + f.s1.x.v2], int32(t7.s)*int32(t7.s))

// Vectorize
let t7.s.widened.f.s1.rdom$x = int16x32(ramp(((f.s1.x.v1*16) + f.min.0)*2, 1, 32))
f[ramp(f.s1.x.v1*16, 1, 16) aligned(16, 0)] = (int32x16)saturating_add(f[ramp(f.s1.x.v1*16, 1, 16) aligned(16, 0)], (int32x16)vector_reduce(SaturatingAdd, (int32x32((int16x32)t7.s.widened.f.s1.rdom$x)*int32x32((int16x32)t7.s.widened.f.s1.rdom$x))))
```

A quick solution is to repeat the find_intrisics pass after vectorize_loops, but I was wondering if you could give me any hints on what to look to fix this and keep the find_intrisics before unrolling

Reply:
As I suspected, just moving the find_intrinsics pass will break the tests, not only for the newly added saturating dot product, but for other intrinsics as well. 

I guess both of my proposed solutions work but I would like to know your thoughts as well @dsharletg

Reply:
It looks like VectorizeLoops.cpp looks for particular expressions in quite a few places that would also need to look for intrinsics, for example here: https://github.com/halide/Halide/blob/master/src/VectorizeLoops.cpp#L1067

I don't think we should add new overloads to the list of patterns. That's basically just replicating the logic in find_intrinsics, which is non-trivial.

Reply:
@dsharletg  I rebased and reverted the last two commits to avoid the failure in the tests. I understand your reasoning for moving the find_instrinsics pass earlier in the pipeline but that doesn't seem to work well.